src/main/java/org/jabref/logic/importer/FulltextFetchers.java

```diff
@@ -36,26 +48,93 @@ public FulltextFetchers(ImportFormatPreferences importFormatPreferences) {
         Optional<DOI> doi = clonedEntry.getField(FieldName.DOI).flatMap(DOI::parse);
 
         if (!doi.isPresent()) {
+            findDoiForEntry(clonedEntry);
+        }
+
+        ExecutorService executor = Executors.newCachedThreadPool();
+        try {
+            List<Future<Optional<FetcherResult>>> result = new ArrayList<>();
             try {
-                WebFetchers.getIdFetcherForIdentifier(DOI.class)
-                        .findIdentifier(clonedEntry)
-                        .ifPresent(e -> clonedEntry.setField(FieldName.DOI, e.getDOI()));
-            } catch (FetcherException e) {
-                LOGGER.debug("Failed to find DOI", e);
+                result = executor.invokeAll(getCallables(clonedEntry, finders), FETCHER_TIMEOUT, TimeUnit.SECONDS);
+            } catch (InterruptedException ignore) {
+
             }
+
+            return result.stream()
+                    .map(FulltextFetchers::waitForResult)
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .filter(res -> Objects.nonNull(res.source))
+                    .sorted(Comparator.comparingInt((FetcherResult res) -> res.trust.getTrustScore()).reversed())
+                    .map(res -> res.source)
+                    .findFirst();
+        } finally {
+            shutdownAndAwaitTermination(executor);
         }
+    }
 
-        for (FulltextFetcher finder : finders) {
+    private void findDoiForEntry(BibEntry clonedEntry) {
+        try {
+            WebFetchers.getIdFetcherForIdentifier(DOI.class)
+                    .findIdentifier(clonedEntry)
+                    .ifPresent(e -> clonedEntry.setField(FieldName.DOI, e.getDOI()));
+        } catch (FetcherException e) {
+            LOGGER.debug("Failed to find DOI", e);
+        }
+    }
+
+    private static Optional<FetcherResult> waitForResult(Future<Optional<FetcherResult>> future) {
+        while (true) {
             try {
-                Optional<URL> result = finder.findFullText(clonedEntry);
+                return future.get();
+            } catch (InterruptedException retry) {
+
+            } catch (ExecutionException | CancellationException e) {
+                LOGGER.debug("Fetcher execution failed or was cancelled");
+                return Optional.empty();
+            }
+        }
+    }
+
+    private Callable<Optional<FetcherResult>> getCallable(BibEntry entry, FulltextFetcher fetcher) {
+            return () -> {
+                try {
+                    Optional<URL> result = fetcher.findFullText(entry);
+
+                    if (result.isPresent() && new URLDownload(result.get().toString()).isPdf()) {
```

Comment:
Do we really need to check that the URL returned is a PDF? This is an additional request. I think, we should just trust the fetcher that it found an appropriate document.

Reply:
This is the single point where this is checked. Otherwise every fetcher has to do this logic. It is vital so we don't get wrong fulltexts.

Reply:
Yes, I was wondering why this check is necessary at all. If the fetcher determines that a given URL contains the right full-text (say because the API tell it so), then why do we need to check that it is a PDF?

Reply:
Because sometimes it is an HTML site or the access is blocked for the users, resulting in something else than the fulltext 😄